nexus/deployment/src/planner.rs

```diff
@@ -0,0 +1,227 @@
+                    );
+                    ncrucibles_added += 1;
+                }
+            }
+
+            if ncrucibles_added > 0 {
+                // Don't make any other changes to this sled.  However, this
+                // change is compatible with any other changes to other sleds,
+                // so we can "continue" here rather than "break".
+                // (Yes, it's currently the last thing in the loop, but being
+                // explicit here means we won't forget to do this when more code
+                // is added below.)
+                self.blueprint.comment(&format!("sled {}: add zones", sled_id));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::Planner;
+    use crate::blueprint_builder::test::example;
+    use crate::blueprint_builder::test::policy_add_sled;
+    use crate::blueprint_builder::BlueprintBuilder;
+    use omicron_common::api::external::Generation;
+    use omicron_test_utils::dev::test_setup_log;
+    use sled_agent_client::types::OmicronZoneType;
+
+    /// Runs through a basic sequence of blueprints for adding a sled
+    #[test]
+    fn test_basic_add_sled() {
+        let logctx = test_setup_log("planner_basic_add_sled");
+
+        // Use our example inventory collection.
+        let (collection, mut policy) = example();
+
+        // Build the initial blueprint.  We don't bother verifying it here
+        // because there's a separate test for that.
+        let blueprint1 = BlueprintBuilder::build_initial_from_collection(
+            &collection,
+            &policy,
+            "the_test",
+        )
+        .expect("failed to create initial blueprint");
+
+        // Now run the planner.  It should do nothing because our initial
+        // system didn't have any issues that the planner currently knows how to
+        // fix.
+        let blueprint2 = Planner::new_based_on(
+            logctx.log.clone(),
+            &blueprint1,
+            &policy,
+            "no-op?",
+        )
+        .plan()
+        .expect("failed to plan");
+
+        let diff = blueprint1.diff(&blueprint2);
+        println!("1 -> 2 (expected no changes):\n{}", diff);
+        assert_eq!(diff.sleds_added().count(), 0);
+        assert_eq!(diff.sleds_removed().count(), 0);
+        assert_eq!(diff.sleds_changed().count(), 0);
+
+        // Now add a new sled.
+        let new_sled_id =
+            "7097f5b3-5896-4fff-bd97-63a9a69563a9".parse().unwrap();
+        let _ = policy_add_sled(&mut policy, new_sled_id);
+
+        // Check that the first step is to add an NTP zone
+        let blueprint3 = Planner::new_based_on(
+            logctx.log.clone(),
+            &blueprint2,
+            &policy,
+            "test: add NTP?",
+        )
+        .plan()
+        .expect("failed to plan");
+
+        let diff = blueprint2.diff(&blueprint3);
+        println!("2 -> 3 (expect new NTP zone on new sled):\n{}", diff,);
```

Comment:
```suggestion
        println!("2 -> 3 (expect new NTP zone on new sled):\n{}", diff);
```

Reply:
Fixed in c58c7edae525570e703cb113075211ccd9903c6c.